pkg/didcomm/protocol/introduce/service.go

```diff
@@ -26,28 +31,173 @@ const (
+type message struct {
+	Msg           *service.DIDCommMsg
+	NextStateName string
+}
+
+var logger = log.New("aries-framework/introduce/service")
+
 // Service for introduce protocol
 type Service struct {
 	service.Action
 	service.Message
+	store storage.Store
 }
 
 // New returns introduce service
-func New() *Service {
-	return &Service{}
+func New(s storage.Provider) (*Service, error) {
+	store, err := s.OpenStore(Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: storage should be closed e.g s.CloseStore(Introduce)
+	return &Service{store: store}, nil
 }
 
-// Handle didexchange msg
+// Handle DIDCommMsg (introduce protocol)
 func (s *Service) Handle(msg *service.DIDCommMsg) error {
+	aEvent := s.GetActionEvent()
+
+	logger.Infof("entered into Handle: %v", msg.Header)
+	// throw error if there is no action event registered for inbound messages
+	if !msg.Outbound && aEvent == nil {
+		return errors.New("no clients are registered to handle the message")
+	}
+
+	thID, err := msg.ThreadID()
+	if err != nil {
+		return err
+	}
+	logger.Infof("thread id value for the message: %s", thID)
+
+	current, err := s.currentState(thID)
+	if err != nil {
+		return err
+	}
+	logger.Infof("current state : %s", current.Name())
+
+	next, err := stateFromMsgType(msg)
+	if err != nil {
+		return err
+	}
+
+	logger.Infof("state will transition from %q to %q if the msgType is processed", current.Name(), next.Name())
+
+	if !current.CanTransitionTo(next) {
+		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
+	}
+
+	// trigger message events
+	s.sendMsgEvents(service.StateMsg{
+		Type:         service.PreState,
+		Msg:          msg,
+		ProtocolName: Introduce,
+		StateID:      next.Name(),
+		Properties:   nil,
+	})
+
+	logger.Infof("sent pre event for state %s", next.Name())
+
+	// trigger action event based on message type for inbound messages
+	if canTriggerActionEvents(msg) {
+		return s.sendActionEvent(msg, aEvent, next)
+	}
+
+	// if no action event is triggered, continue the execution
+	return s.handle(message{
+		Msg:           msg,
+		NextStateName: next.Name(),
+	})
+}
+
+// sendMsgEvents triggers the message events.
+func (s *Service) sendMsgEvents(msg service.StateMsg) {
+	// trigger the message events
+	for _, handler := range s.GetMsgEvents() {
+		handler <- msg
+	}
+}
+
+// sendEvent triggers the action event. This function stores the state of current processing and passes a callback
+// function in the event message.
+func (s *Service) sendActionEvent(msg *service.DIDCommMsg, aEvent chan<- service.DIDCommAction, nextState state) error {
 	return nil
 }
 
+func stateFromMsgType(msg *service.DIDCommMsg) (state, error) {
+	switch msg.Header.Type {
+	case ProposalMsgType:
+		if msg.Outbound {
+			return &deciding{}, nil
+		}
+		return &arranging{}, nil
+	default:
+		return nil, fmt.Errorf("unrecognized msgType: %s", msg.Header.Type)
```

Comment:
I assume this default behaviour is temporary and that subsequent PRs will add cases for the other message types. Confirm?

Reply:
yeah, I did only one to cover the test case scenario. But the logic here is correct 
```
if msg.Outbound {
   return &deciding{}, nil
}
return &arranging{}, nil
```
also, it will depend on `wait_count` as well